diff --git a/index.d.ts b/index.d.ts
index 9b333ef08adedb70487fac220b7d71d2592efaa8..b107a59db262cf533b3c29c72c451270e8ba652e 100644
--- a/index.d.ts
+++ b/index.d.ts
@@ -1,6 +1,6 @@
-export * from "./libs/ast-types";
-export * from "./libs/type-guard";
-export * from "./libs/info-specs";
+export * from "./libs/ast-types.js";
+export * from "./libs/type-guard.js";
+export * from "./libs/info-specs.js";
 /**
  * ## What is this?
  *
diff --git a/libs/ast-types.d.ts b/libs/ast-types.d.ts
index c65aca15bf97255cce6fb40403e741088ba8dae2..b7f940c6299aec7eda0d25ab42092427191b3ba1 100644
--- a/libs/ast-types.d.ts
+++ b/libs/ast-types.d.ts
@@ -1,4 +1,4 @@
-import { EnvInfo, MacroInfo } from "./info-specs";
+import { EnvInfo, MacroInfo } from "./info-specs.js";
 export type GenericAst = GenericNode | GenericNode[];
 export interface GenericNode {
     [x: string]: any;
diff --git a/libs/info-specs.d.ts b/libs/info-specs.d.ts
index 1a09e416a4ef036f4c75cf3dbee43c85774e7060..2a0e78a0d39510a60804025d013dfe5942d23cd4 100644
--- a/libs/info-specs.d.ts
+++ b/libs/info-specs.d.ts
@@ -1,4 +1,4 @@
-import * as Ast from "./ast-types";
+import * as Ast from "./ast-types.js";
 export type ArgumentParser = (nodes: Ast.Node[], startPos: number) => {
     args: Ast.Argument[];
     nodesRemoved: number;
